### PR TITLE
fix: use url-join to properly support Travis Enterprise

### PR DESCRIPTION
**Pull request by SimenB**
Closes https://github.com/semantic-release/condition-travis/pull/105

This feels like a better fix

### CODE REVIEW

**Comment by SimenB** [comment]
Don't merge this, it doesn't work (yet)...

**Comment by SimenB** [comment]
Never mind it works, I was fooled by travis caching.

File: lib/get-travis-url.js [hunk 1]
```diff
@@ -1,2 +1,9 @@
-module.exports = ({enterprise, pro} = {}) =>
-  enterprise || (pro ? 'https://api.travis-ci.com' : 'https://api.travis-ci.org');
+const url = require('url');
+
+module.exports = ({enterprise, pro} = {}) => {
+  if (enterprise) {
+    return url.resolve(enterprise, 'api');
```
Comment (SimenB) [inline lib/get-travis-url.js]: it's always `/api`.

https://docs.travis-ci.com/api/

>Travis Enterprise: For projects running on a custom setup, use https://travis.example.com/api (where you replace travis.example.com with the domain Travis CI is running on).

**Comment by pvdlg** [comment]
Regarding your mention of `githubUrl` and `githubApiPathPrefix` in semantic-release/condition-travis#105. We do that for Github as it's the way the API client works. In addition we are using the `githubUrl` to determine other things than the API URL (likes the issues, PR, commit, asset upload URL etc).

With Travis we don't do that and we use only the API URL.

Even if the doc mention the API is always in the `/api` context, I imagine the Travis Enterprise instance can be behind a reverse proxy and be exposed in a different path. Or it can be deployed in a different context (like `/custom/api`).

So I'm not sure adding the `/api` context might create some limitation in some specific cases. Does adding it brings any value? What's the difference for the user between setting `http://custom.url/api` vs `http://custom.url` in the environment variable / options?

Maybe it would be better to keep it as is and just mention in the documentation that the expected URL in `enterprise` is the complete API endpoint?

**Comment by SimenB** [comment]
We cannot keep it as is, as `url.resolve('https://travis.example.com/api', 'auth/github')` returns `https://travis.example.com/auth/github'`.

**Comment by SimenB** [comment]
so either we need path explicitly passed in, or we need to join urls in some other way than `url.resolve`

**Comment by pvdlg** [comment]
> We cannot keep it as is, as url.resolve('https://travis.example.com/api', 'auth/github') returns https://travis.example.com/auth/github' (first argument to resolve is stripped of its path).

Indeed. My mistake.
This
```js
return (await got(url.resolve(getTravisUrl(travisOpts), `builds/${buildId}`), {
```
should be
```js
return (await got(url.resolve(getTravisUrl(travisOpts), `/builds/${buildId}`), {
```
There must be a `/` in front of `builds/${buildId}`.

Could you:
- revert your changes
- add the missing `/` (in two places I think)
- Modify the test `Return job list with Travis Enterprise` to add a prefix at the end of the URL
- Clarify in the doc that the URL expected is the full URL of the API

Thanks!

**Comment by SimenB** [comment]
`url.resolve` always strips the path, adding slashes doesn't help. 🙁 

```sh-session
$ node -p "require('url').resolve('https://travis.example.com/api', '/builds/123')"
https://travis.example.com/builds/123
node -p "require('url').resolve('https://travis.example.com/api/', '/builds/123')"
https://travis.example.com/builds/123
```

**Comment by pvdlg** [comment]
Indeed, I misread the doc. Somehow, it's recommended in many different places to use `resolve` to concat URL, but it seems to really not be a good idea.

What do you think about the following?
```
// get-travis-url.js

module.exports = ({enterprise, pro} = {}) =>
  (enterprise || (pro ? 'https://api.travis-ci.com' : 'https://api.travis-ci.org')).replace(/[^/]$/, '$&/')
```
```js
// get-jobs and and get-token.js

await got(`${getTravisUrl(travisOpts)}builds/${buildId}`), {...}
```

**Comment by SimenB** [comment]
That feels convoluted, what about just adding https://www.npmjs.com/package/url-join? Seems to solve the problem pretty good

**Comment by pvdlg** [comment]
Ok. It's crazy there is no standard Node API do that...It's a really common thing to do...

**Comment by SimenB** [comment]
Pushed 🙂

File: lib/get-jobs.js [hunk 2]
```diff
@@ -25,7 +25,7 @@ module.exports = async (travisOpts, travisToken, buildId) => {
   } catch (err) {
     // https://github.com/semantic-release/travis-deploy-once/issues/3
     // https://github.com/pwmckenna/node-travis-ci/issues/17
-    if (err.response.body && err.response.body.file === 'not found') {
+    if (err.response && err.response.body && err.response.body.file === 'not found') {
```
Comment (SimenB) [inline lib/get-jobs.js]: I messed up in the test (got `undefined` in the host, so nock didn't match), and got this:
image

File: test/helpers/mock-travis.js [hunk 3]
```diff
@@ -2,18 +2,24 @@ import nock from 'nock';
 
 /* eslint camelcase: ["error", {properties: "never"}] */
 
+function getBaseUrl(enterprise, pro) {
```
Comment (pvdlg) [inline test/helpers/mock-travis.js]: All the changes in `mock-travis.js` are not necessary. With the current code `nock` will register with whatever is in `interprise` including the `/api`.
Just don't pass the `/api` twice in the tests.

File: test/get-jobs.test.js [hunk 4]
```diff
@@ -50,14 +50,14 @@ test.serial('Return job list with Travis Pro', async t => {
...
 test.serial('Return job list with Travis Enterprise', async t => {
   const travisToken = 'TRAVIS_TOKEN';
-  const travisOpts = {pro: false, enterprise: 'https://travis.example.com'};
+  const travisOpts = {pro: false, enterprise: 'https://travis.example.com/api'};
   const buildId = 123;
   const jobsFirst = [{id: 456, state: 'started'}, {id: 789, state: 'started'}];
   const jobsSecond = [{id: 456, state: 'passed'}, {id: 789, state: 'passed'}];
   const travis = api({travisOpts, travisToken})
-    .get(`/builds/${buildId}`)
+    .get(`/api/builds/${buildId}`)
```
Comment (pvdlg) [inline test/get-jobs.test.js]: No need to add `api/` here, it's already in `travisOpts.enterprise `

File: test/get-jobs.test.js [hunk 5]
```diff
@@ -50,14 +50,14 @@ test.serial('Return job list with Travis Pro', async t => {
...
     .reply(200, {jobs: jobsFirst})
-    .get(`/builds/${buildId}`)
+    .get(`/api/builds/${buildId}`)
```
Comment (pvdlg) [inline test/get-jobs.test.js]: No need to add `api/` here, it's already in `travisOpts.enterprise `

File: README.md [hunk 6]
```diff
@@ -80,4 +80,4 @@ Default: `false`
 
 Type: `String`
 
-[Travis Enterprise](https://enterprise.travis-ci.com) URL. If defined, the [pro](#pro) option will be ignored.
+[Travis Enterprise](https://enterprise.travis-ci.com) URL. If defined, the [pro](#pro) option will be ignored. Note that this has to be the full URL, e.g. `https://travis.example.com/api`.
```
Comment (pvdlg) [inline README.md]: Maybe something along those lines that to make it more clear:
```
[Travis Enterprise](https://enterprise.travis-ci.com) URL. If defined, the [pro](#pro) option will be ignored.

**Note**: This is the URL of the API endpoint, for example `https://travis.example.com/api`.
```

Comment (SimenB) [reply README.md]: Done

**Comment by SimenB** [comment]
Pushed a commit reverting the changes ot the tests.

My thinking was that having the tests tes what did not work before was a good idea, we still keep 100% coverage so 🤷‍♂️

File: test/get-jobs.test.js [hunk 7]
```diff
@@ -50,14 +50,14 @@ test.serial('Return job list with Travis Pro', async t => {
 
 test.serial('Return job list with Travis Enterprise', async t => {
   const travisToken = 'TRAVIS_TOKEN';
-  const travisOpts = {pro: false, enterprise: 'https://travis.example.com'};
+  const travisOpts = {pro: false, enterprise: 'https://travis.example.com/api'};
```
Comment (pvdlg) [inline test/get-jobs.test.js]: The `/api` should stay here.

File: test/get-token.test.js [hunk 8]
```diff
@@ -57,7 +57,7 @@ test.serial('Authenticate with Travis Pro', async t => {
 
 test.serial('Authenticate with Travis Enterprise', async t => {
   process.env.GH_TOKEN = 'GITHUB_TOKEN';
-  const travisOpts = {pro: true, enterprise: 'https://travis.example.com'};
+  const travisOpts = {pro: true, enterprise: 'https://travis.example.com/api'};
```
Comment (pvdlg) [inline test/get-token.test.js]: The `/api` should stay here.

File: test/get-jobs.test.js (hunk 7 above)
Comment (SimenB) [reply test/get-jobs.test.js]: Ah, OK! I misunderstood your requested change then.

**Comment by pvdlg** [comment]
> My thinking was that having the tests test what did not work before was a good idea, but we still keep 100% coverage without changing tests, so 🤷‍♂️

Yes we should do that by testing with `/api` in `test/get-jobs.test.js` and `test/get-token.test.js`.

See [comment](https://github.com/semantic-release/travis-deploy-once/pull/32#discussion_r156786591) and [comment](https://github.com/semantic-release/travis-deploy-once/pull/32#discussion_r156786656)

**Comment by SimenB** [comment]
Perfect, I misunderstood the API of nock, I though `nock(url)` could just be the host, not path. Re-added `/api` now, and can verify that with `url.resolve` the two tests fail. Thanks!

**Comment by SimenB** [comment]
Yay! I'll test it out tomorrow at work and report back :)

**Pull request merged**